ffs-shared/rule/src/commonTest/kotlin/rule/RuleEvalTest.kt

```diff
@@ -281,6 +281,27 @@ class RuleEvalTest {
         assertFailsWith<IllegalArgumentException> { eval("""exp("1")""") }
     }
 
+    @Test
+    fun ipv4matches() {
+        assertEquals(0f, eval("""ipv4matches("10.0.0.0/28", "10.0.0.16")"""))
+        assertEquals(0f, eval("""ipv4matches("10.0.0.0/28", "9.255.255.255")"""))
+        assertEquals(0f, eval("""ipv4matches("192.168.2.10/24", "192.168.3.255")"""))
+        assertEquals(1f, eval("""ipv4matches("10.0.0.0/28", "10.0.0.15")"""))
+        assertEquals(1f, eval("""ipv4matches("255.0.0.0/28", "255.0.0.4")"""))
+        assertEquals(1f, eval("""ipv4matches("192.168.2.10/24", "192.168.2.254")"""))
```

Comment:
> **Ergonomics** Could it be ergonomic and flexible to leverage the existing composability of the existing functions, instead of having multiple `match` variants? For instance:
> 
> * `contains(cidr("10.0.0.0/28"), ip("10.0.0.16"))`
> * `lt(ip(env["user.ip"], ip("109.71.47.255"))`
> * `not(ip(env["user.ip"], ip("127.0.0.1"))`
> 
> Of course `ip` and `cidr` would have to return something. `Long` and `Range` could be good choices, but this is related to the next point.

I agree, I haven't put more thought into this, and creating the rule to be composable seems to be a lot better. So there will be 2 new function

cidr -> return `Range`
ip -> return `Long`

> **Performance** The current proposal relies on a non-trivial regular expression that we can't pre-compile, which can be suboptimal considering FFS is designed to do thousands of evaluations per second. But IP addresses are just numbers, and IP+mask are just ranges. We could treat them like that?

In this case, the regex is redundant, it was only used to validate the IP / CIDR format. Referring to the Ergonomics, treating IP addresses and mask as numbers are to be expected.

> Specifically, what if we:
> 
> * Add a [range type](https://gist.github.com/goncalossilva/2406eb278eb0c8a1b89c3db76f4e0e2b) (with tests & docs). It seems useful, regardless of IPs.
> * Implement `ip(value)` as a function that takes an IP, and returns a `Long`. An IP is just a group of 4 bytes, so we could parse it ([example](https://stackoverflow.com/questions/62124452/converting-cidr-notation-into-ip-range-without-other-library)) and combine it together in a single number.
> * Implement `cidr(value)` as a function that takes an IP+mast, and returns a range. It can reuse the range and `ip(value)` implementations from before.
> 
> This could be very efficient. What do you think?

Yes, this will be efficient to run. My only concern is the IPv6 format. It's not straightforward to convert the IP to number, even using ULong, so we'll need to find a way to represent the IPv6 in 1 single variable while keeping it comparable to the CIDR

For the Alpha milestone, will it be enough if we only support IPv4?